### DISCUSSION TITLE
[Question][Hw5] Variable of type llvalue

### DISCUSSION THREAD

**Discussion by pacokwon** [prosyslab-classroom/cs348-information-security#591] [Q&A]
In `domain.ml`'s `module Variable` declaration below,
```ocaml
module Variable : SET with type t = Llvm.llvalue = struct
  type t = Llvm.llvalue

  let compare = compare

  let pp fmt v = Utils.string_of_lhs v |> Format.fprintf fmt "%s"
end

```
I see that the `Variable` module in `domain.ml` has type `t` as `llvalue`,  and the `pp` function uses `Utils.string_of_lhs` on this. I am confused on this concept and would appreciate some advice:

1. What kind of `llvalue` is used to represent a variable? For example, which utility function in `Llvm` or `Utils` can I refer to, to extract a form of `llvalue` that represents a variable?
2. The `Utils.string_of_lhs` function seems to use regex on `" = "` and extract the lhs. From what I can infer, this implies that `Llvm.string_of_llvalue` of this `v` should contain a string `" = "`, which then makes it seem like `v` should contain an `llvalue` that represents an assignment (and not a variable, which only corresponds to the lhs of the assignment).

(Question 2 is a bit messy, I'm sorry)

**Comment by KihongHeo** [comment]
1. `llvalue` itself is the datatype for variables. See this [document](https://github.com/prosyslab-classroom/llvm-primer/blob/master/data-structure.md).
2. What is your question? `string_of_lhs` is just for pretty printer for debugging.

**Comment by pacokwon** [comment]
Seeing the document, I think I understand it now. Thank you professor!
